File: swift/Sources/IdkollenClient/BankIdNo.swift
import Foundation

// MARK: - Requests

public struct BankIdNoAuthRequest: Encodable {
    public let redirectUrl: String?
    public let requestSsn: Bool?
    public let requestPhone: Bool?
    public let requestEmail: Bool?
    public let requestAddress: Bool?
    public let refId: String?
    public let appCallbackUri: String?

    public init(
        redirectUrl: String? = nil,
        requestSsn: Bool? = nil,
        requestPhone: Bool? = nil,
        requestEmail: Bool? = nil,
        requestAddress: Bool? = nil,
        refId: String? = nil,
        appCallbackUri: String? = nil
    ) {
        self.redirectUrl = redirectUrl
        self.requestSsn = requestSsn
        self.requestPhone = requestPhone
        self.requestEmail = requestEmail
        self.requestAddress = requestAddress
        self.refId = refId
        self.appCallbackUri = appCallbackUri
    }
}

public struct BankIdNoBackchannelAuthRequest: Encodable {
    public let ssn: String
    public let callbackUrl: String?
    public let refId: String?

    public init(ssn: String, callbackUrl: String? = nil, refId: String? = nil) {
        self.ssn = ssn
        self.callbackUrl = callbackUrl
        self.refId = refId
    }
}

public struct BankIdNoSignRequest: Encodable {
    public let redirectUrl: String?
    public let text: String?
    public let documents: [String]?
    public let requestSsn: Bool?
    public let requestPhone: Bool?
    public let requestEmail: Bool?
    public let requestAddress: Bool?
    public let refId: String?

    public init(
        redirectUrl: String? = nil,
        text: String? = nil,
        documents: [String]? = nil,
        requestSsn: Bool? = nil,
        requestPhone: Bool? = nil,
        requestEmail: Bool? = nil,
        requestAddress: Bool? = nil,
        refId: String? = nil
    ) {
        self.redirectUrl = redirectUrl
        self.text = text
        self.documents = documents
        self.requestSsn = requestSsn
        self.requestPhone = requestPhone
        self.requestEmail = requestEmail
        self.requestAddress = requestAddress
        self.refId = refId
    }
}

// MARK: - Helper structs

public struct BankIdNoSignResult: Decodable {
    public let endUser: String
    public let merchant: String
    public let hash: String
}

public struct BankIdNoSignedDocument: Decodable {
    public let id: String
    public let hash: String
}

// MARK: - Status enum + variants

public enum BankIdNoStatus: Decodable {
    case pending(BankIdNoPending)
    case completed(BankIdNoCompleted)
    case failed(BankIdNoFailed)

    private enum DiscriminatorKey: String, CodingKey { case status }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: DiscriminatorKey.self)
        let status = try container.decode(String.self, forKey: .status)
        switch status {
        case "PENDING":   self = .pending(try BankIdNoPending(from: decoder))
        case "COMPLETED": self = .completed(try BankIdNoCompleted(from: decoder))
        case "FAILED":    self = .failed(try BankIdNoFailed(from: decoder))
        default:
            throw DecodingError.dataCorruptedError(forKey: .status, in: container, debugDescription: "Unknown bankid-no status: \(status)")
        }
    }
}

public struct BankIdNoPending: Decodable {
    public let id: String
    public let refId: String?
    public let url: String?
    public let bindingMessage: String?
}

public struct BankIdNoCompleted: Decodable {
    public let id: String
    public let refId: String?
    public let ssn: String
    public let name: String
    public let givenName: String
    public let surname: String
    public let phone: String?
    public let email: String?
    public let address: String?
    public let birthDate: String?
    public let pid: String?
    public let bankId: String?
    public let signResult: BankIdNoSignResult?
    public let signedDocuments: [BankIdNoSignedDocument]?
}

public struct BankIdNoFailed: Decodable {
    public let id: String
    public let refId: String?
    public let error: String
}

// MARK: - Endpoint

public final class BankIdNoEndpoint {
    private let transport: Transport
    internal init(_ transport: Transport) { self.transport = transport }

    public func auth(_ req: BankIdNoAuthRequest) async throws -> BankIdNoStatus {
        try await transport.post("/v3/bankid-no/auth", body: req)
    }

    public func backchannelAuth(_ req: BankIdNoBackchannelAuthRequest) async throws -> BankIdNoStatus {
        try await transport.post("/v3/bankid-no/backchannel/auth", body: req)
    }

    public func sign(_ req: BankIdNoSignRequest) async throws -> BankIdNoStatus {
        try await transport.post("/v3/bankid-no/sign", body: req)
    }

    public func authStatus(id: String) async throws -> BankIdNoStatus {
        try await transport.get("/v3/bankid-no/auth/\(id)")
    }

    public func signStatus(id: String) async throws -> BankIdNoStatus {
        try await transport.get("/v3/bankid-no/sign/\(id)")
    }

    public func cancelAuth(id: String) async throws {
        try await transport.delete("/v3/bankid-no/auth/\(id)")
    }

    public func cancelSign(id: String) async throws {
        try await transport.delete("/v3/bankid-no/sign/\(id)")
    }

    public func waitForAuth(id: String, opts: PollOptions = .init()) async throws -> BankIdNoStatus {
        try await poll({ try await self.authStatus(id: id) }, opts: opts)
    }

    public func ageVerification(_ req: AgeVerificationRequest) async throws -> AgeVerificationStatus {
        try await transport.post("/v3/bankid-no/age-verification", body: req)
    }

    public func ageVerificationStatus(id: String) async throws -> AgeVerificationStatus {
        try await transport.get("/v3/bankid-no/age-verification/\(id)")
    }

    public func cancelAgeVerification(id: String) async throws {
        try await transport.delete("/v3/bankid-no/age-verification/\(id)")
    }

    public func waitForAgeVerification(id: String, opts: PollOptions = .init()) async throws -> AgeVerificationStatus {
        let deadline = Date().addingTimeInterval(opts.timeout)
        while true {
            let status = try await ageVerificationStatus(id: id)
            if case .pending = status {
                if Date() >= deadline { throw WaitError(timeout: true) }
                try await Task.sleep(nanoseconds: UInt64(opts.interval * 1_000_000_000))
            } else {
                return status
            }
        }
    }

    public func waitForSign(id: String, opts: PollOptions = .init()) async throws -> BankIdNoStatus {
        try await poll({ try await self.signStatus(id: id) }, opts: opts)
    }

    private func poll(_ fn: @escaping () async throws -> BankIdNoStatus, opts: PollOptions) async throws -> BankIdNoStatus {
        let deadline = Date().addingTimeInterval(opts.timeout)
        while true {
            let status = try await fn()
            if case .pending = status {
                if Date() >= deadline { throw WaitError(timeout: true) }
                try await Task.sleep(nanoseconds: UInt64(opts.interval * 1_000_000_000))
            } else {
                return status
            }
        }
    }
}

File: dart/lib/src/bankid_no.dart
import 'exceptions.dart';
import 'poll_options.dart';
import 'transport.dart';

class BankIdNoAuthRequest {
  const BankIdNoAuthRequest({
    this.redirectUrl,
    this.requestSsn,
    this.requestPhone,
    this.requestEmail,
    this.requestAddress,
    this.refId,
    this.appCallbackUri,
  });

  final String? redirectUrl;
  final bool? requestSsn;
  final bool? requestPhone;
  final bool? requestEmail;
  final bool? requestAddress;
  final String? refId;
  final String? appCallbackUri;

  Map<String, dynamic> toJson() => {
        if (redirectUrl != null) 'redirectUrl': redirectUrl,
        if (requestSsn != null) 'requestSsn': requestSsn,
        if (requestPhone != null) 'requestPhone': requestPhone,
        if (requestEmail != null) 'requestEmail': requestEmail,
        if (requestAddress != null) 'requestAddress': requestAddress,
        if (refId != null) 'refId': refId,
        if (appCallbackUri != null) 'appCallbackUri': appCallbackUri,
      };
}

class BankIdNoBackchannelAuthRequest {
  const BankIdNoBackchannelAuthRequest({required this.ssn, this.callbackUrl, this.refId});
  final String ssn;
  final String? callbackUrl;
  final String? refId;
  Map<String, dynamic> toJson() => {
        'ssn': ssn,
        if (callbackUrl != null) 'callbackUrl': callbackUrl,
        if (refId != null) 'refId': refId,
      };
}

class BankIdNoSignRequest {
  const BankIdNoSignRequest({
    this.redirectUrl,
    this.text,
    this.documents,
    this.requestSsn,
    this.requestPhone,
    this.requestEmail,
    this.requestAddress,
    this.refId,
  });

  final String? redirectUrl;
  final String? text;
  final List<String>? documents;
  final bool? requestSsn;
  final bool? requestPhone;
  final bool? requestEmail;
  final bool? requestAddress;
  final String? refId;

  Map<String, dynamic> toJson() => {
        if (redirectUrl != null) 'redirectUrl': redirectUrl,
        if (text != null) 'text': text,
        if (documents != null) 'documents': documents,
        if (requestSsn != null) 'requestSsn': requestSsn,
        if (requestPhone != null) 'requestPhone': requestPhone,
        if (requestEmail != null) 'requestEmail': requestEmail,
        if (requestAddress != null) 'requestAddress': requestAddress,
        if (refId != null) 'refId': refId,
      };
}

class BankIdNoSignResult {
  const BankIdNoSignResult({required this.endUser, required this.merchant, required this.hash});
  final String endUser;
  final String merchant;
  final String hash;
  factory BankIdNoSignResult.fromJson(Map<String, dynamic> json) => BankIdNoSignResult(
        endUser: json['endUser'] as String,
        merchant: json['merchant'] as String,
        hash: json['hash'] as String,
      );
}

class BankIdNoSignedDocument {
  const BankIdNoSignedDocument({required this.id, required this.hash});
  final String id;
  final String hash;
  factory BankIdNoSignedDocument.fromJson(Map<String, dynamic> json) => BankIdNoSignedDocument(
        id: json['id'] as String,
        hash: json['hash'] as String,
      );
}

sealed class BankIdNoStatus {
  factory BankIdNoStatus.fromJson(Map<String, dynamic> json) => switch (json['status']) {
        'PENDING' => BankIdNoPending.fromJson(json),
        'COMPLETED' => BankIdNoCompleted.fromJson(json),
        'FAILED' => BankIdNoFailed.fromJson(json),
        _ => throw FormatException('Unknown bankid-no status: ${json['status']}'),
      };
}

final class BankIdNoPending implements BankIdNoStatus {
  const BankIdNoPending({required this.id, this.refId, this.url, this.bindingMessage});
  final String id;
  final String? refId;
  final String? url;
  final String? bindingMessage;
  factory BankIdNoPending.fromJson(Map<String, dynamic> json) => BankIdNoPending(
        id: json['id'] as String,
        refId: json['refId'] as String?,
        url: json['url'] as String?,
        bindingMessage: json['bindingMessage'] as String?,
      );
}

final class BankIdNoCompleted implements BankIdNoStatus {
  const BankIdNoCompleted({
    required this.id,
    this.refId,
    required this.ssn,
    required this.name,
    required this.givenName,
    required this.surname,
    this.phone,
    this.email,
    this.address,
    this.birthDate,
    this.pid,
    this.bankId,
    this.signResult,
    this.signedDocuments,
  });

  final String id;
  final String? refId;
  final String ssn;
  final String name;
  final String givenName;
  final String surname;
  final String? phone;
  final String? email;
  final String? address;
  final String? birthDate;
  final String? pid;
  final String? bankId;
  final BankIdNoSignResult? signResult;
  final List<BankIdNoSignedDocument>? signedDocuments;

  factory BankIdNoCompleted.fromJson(Map<String, dynamic> json) => BankIdNoCompleted(
        id: json['id'] as String,
        refId: json['refId'] as String?,
        ssn: json['ssn'] as String,
        name: json['name'] as String,
        givenName: json['givenName'] as String,
        surname: json['surname'] as String,
        phone: json['phone'] as String?,
        email: json['email'] as String?,
        address: json['address'] as String?,
        birthDate: json['birthDate'] as String?,
        pid: json['pid'] as String?,
        bankId: json['bankId'] as String?,
        signResult: json['signResult'] == null
            ? null
            : BankIdNoSignResult.fromJson(json['signResult'] as Map<String, dynamic>),
        signedDocuments: (json['signedDocuments'] as List<dynamic>?)
            ?.map((e) => BankIdNoSignedDocument.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

final class BankIdNoFailed implements BankIdNoStatus {
  const BankIdNoFailed({required this.id, this.refId, required this.error});
  final String id;
  final String? refId;
  final String error;
  factory BankIdNoFailed.fromJson(Map<String, dynamic> json) => BankIdNoFailed(
        id: json['id'] as String,
        refId: json['refId'] as String?,
        error: json['error'] as String,
      );
}

class BankIdNoEndpoint {
  BankIdNoEndpoint(this._transport);
  final Transport _transport;

  Future<BankIdNoStatus> auth(BankIdNoAuthRequest req) async =>
      BankIdNoStatus.fromJson(await _transport.post('/v3/bankid-no/auth', req.toJson()));
  Future<BankIdNoStatus> backchannelAuth(BankIdNoBackchannelAuthRequest req) async =>
      BankIdNoStatus.fromJson(await _transport.post('/v3/bankid-no/backchannel/auth', req.toJson()));
  Future<BankIdNoStatus> sign(BankIdNoSignRequest req) async =>
      BankIdNoStatus.fromJson(await _transport.post('/v3/bankid-no/sign', req.toJson()));
  Future<BankIdNoStatus> authStatus(String id) async =>
      BankIdNoStatus.fromJson(await _transport.get('/v3/bankid-no/auth/$id'));
  Future<BankIdNoStatus> signStatus(String id) async =>
      BankIdNoStatus.fromJson(await _transport.get('/v3/bankid-no/sign/$id'));
  Future<void> cancelAuth(String id) => _transport.delete('/v3/bankid-no/auth/$id');
  Future<void> cancelSign(String id) => _transport.delete('/v3/bankid-no/sign/$id');

  Future<BankIdNoStatus> waitForAuth(String id, {PollOptions opts = const PollOptions()}) =>
      _poll(() => authStatus(id), opts);
  Future<BankIdNoStatus> waitForSign(String id, {PollOptions opts = const PollOptions()}) =>
      _poll(() => signStatus(id), opts);

  Future<BankIdNoStatus> _poll(Future<BankIdNoStatus> Function() fn, PollOptions opts) async {
    final deadline = DateTime.now().add(opts.timeout);
    while (true) {
      final status = await fn();
      if (status is! BankIdNoPending) return status;
      if (DateTime.now().isAfter(deadline)) throw WaitException(timeout: true);
      await Future<void>.delayed(opts.interval);
    }
  }
}
